libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,223 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/ControlFlowRevertPruner.h>
+
+#include <libsolutil/Algorithms.h>
+
+
+namespace solidity::frontend
+{
+
+namespace {
+
+/// Find the right scope for the called function: When calling a base function, we keep the most derived, but we use the called contract in case it is a library function or nullptr for a free function
+ContractDefinition const* findScopeContract(FunctionDefinition const& _function, ContractDefinition const* _callingContract)
+{
+	if (auto const* functionContract = _function.annotation().contract)
+	{
+		if (_callingContract && _callingContract->derivesFrom(*functionContract))
+			return  _callingContract;
+		else
+			return functionContract;
+	}
+
+	return nullptr;
+}
+}
+
+void ControlFlowRevertPruner::run()
+{
+	// build a lookup table for function calls / callers
+	for (auto& [pair, flow]: m_cfg.allFunctionFlows())
+		collectCalls(*pair.function, pair.contract);
+
+	findRevertStates();
+	modifyFunctionFlows();
+}
+
+FunctionDefinition const* ControlFlowRevertPruner::resolveCall(FunctionCall const& _functionCall, ContractDefinition const* _contract)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return nullptr;
+
+	auto const& unresolvedFunctionDefinition =
```

Comment:
Why do you need both `unresolvedFunctionDefinition` and `returnFunctionDef`?

Reply:
mostly for reading clarity. I didn't want to return a re-assigned `unresolvedFunctionDefinition` that is then something else... I can of course just use `returnfunctionDef` for example..